Fix span structure for tracing

ekzhu opened the pull request:
Resolves #5697

jackgerrits commented:
This might mean we can remove some of the tracing metadata that is sent around? Unless there are other uses for it

heyitsaamir reviewed (commented):
I think this makes sense. Have you tried this in practice? I guess all the nesting can happen in the application code. With this do we lose all nesting?

Also, we basically lose any causality here - I wonder if we can retain that even without nested spans...

ekzhu commented:
To test, run this script:

```python
import asyncio
from dataclasses import dataclass
from typing import Callable

from autogen_core import DefaultTopicId, MessageContext, RoutedAgent, default_subscription, message_handler


@dataclass
class Message:
    content: int


@default_subscription
class Modifier(RoutedAgent):
    def __init__(self, modify_val: Callable[[int], int]) -> None:
        super().__init__("A modifier agent.")
        self._modify_val = modify_val

    @message_handler
    async def handle_message(self, message: Message, ctx: MessageContext) -> None:
        val = self._modify_val(message.content)
        print(f"{'-'*80}\nModifier:\nModified {message.content} to {val}")
        await self.publish_message(Message(content=val), DefaultTopicId())  # type: ignore


@default_subscription
class Checker(RoutedAgent):
    def __init__(self, run_until: Callable[[int], bool]) -> None:
        super().__init__("A checker agent.")
        self._run_until = run_until

    @message_handler
    async def handle_message(self, message: Message, ctx: MessageContext) -> None:
        if not self._run_until(message.content):
            print(f"{'-'*80}\nChecker:\n{message.content} passed the check, continue.")
            await self.publish_message(Message(content=message.content), DefaultTopicId())
        else:
            print(f"{'-'*80}\nChecker:\n{message.content} failed the check, stopping.")

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.trace.export import SimpleSpanProcessor
from autogen_core import AgentId, SingleThreadedAgentRuntime

def configure_jaeger_tracing() -> trace.TracerProvider:
    jaeger_exporter = OTLPSpanExporter(endpoint="http://localhost:4317", insecure=True)
    tracer_provider = TracerProvider(
        resource=Resource({"service.name": "autogen-test"})
    )
    span_processor = BatchSpanProcessor(jaeger_exporter)
    tracer_provider.add_span_processor(span_processor)

    return tracer_provider
    
tracer_provider = configure_jaeger_tracing()
tracer = tracer_provider.get_tracer(__name__)
runtime = SingleThreadedAgentRuntime(tracer_provider=tracer_provider)

async def main():
    with tracer.start_as_current_span("runtime"):
        await Modifier.register(
            runtime,
            "modifier",
            # Modify the value by subtracting 1
            lambda: Modifier(modify_val=lambda x: x - 1),
        )

        await Checker.register(
            runtime,
            "checker",
            # Run until the value is less than or equal to 1
            lambda: Checker(run_until=lambda x: x <= 1),
        )

        # Start the runtime and send a direct message to the checker.
        runtime.start()
        await runtime.send_message(Message(3), AgentId("checker", "default"))
        await runtime.stop_when_idle()


asyncio.run(main())
```

output:
```
--------------------------------------------------------------------------------
Checker:
3 passed the check, continue.
--------------------------------------------------------------------------------
Modifier:
Modified 3 to 2
--------------------------------------------------------------------------------
Checker:
2 passed the check, continue.
--------------------------------------------------------------------------------
Modifier:
Modified 2 to 1
--------------------------------------------------------------------------------
Checker:
1 failed the check, stopping.
```

In Jaeger viewer:

<img width="1073" alt="image" src="https://github.com/user-attachments/assets/b10e6618-09af-4cdc-a9a9-09647c69d318" />

ekzhu commented:
@heyitsaamir 
> I think this makes sense. Have you tried this in practice? I guess all the nesting can happen in the application code. With this do we lose all nesting?

Looks like we don't lose nesting. What we get is basically the trace context in the code now starts from each agent's handler rather than from the very first message. So if the application has created a trace context from starting the runtime, it will apply to all agents; If the application starts a trace context from within each agent's handler, it will be within that handler only. 

> Also, we basically lose any causality here - I wonder if we can retain that even without nested spans...

I know, I am not sure how to do it. @TomeHirata @BenWilson2 any thought on how to maintain a causality link between traces? Or is it requires the logs themselves to maintain the causality? E.g., an agent X publishes a message A, and a new trace context is created for handling A in agent Y.

ekzhu commented:
AgentChat is also supported by providing custom `runtime` to the team:

```python
import asyncio

from typing import Sequence
import logging

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.trace.export import SimpleSpanProcessor
from autogen_core import AgentId, SingleThreadedAgentRuntime, EVENT_LOGGER_NAME, TRACE_LOGGER_NAME
from autogen_agentchat.agents import AssistantAgent, UserProxyAgent
from autogen_agentchat.conditions import MaxMessageTermination, TextMentionTermination
from autogen_agentchat.messages import AgentEvent, ChatMessage
from autogen_agentchat.teams import SelectorGroupChat, RoundRobinGroupChat
from autogen_agentchat.ui import Console
from autogen_ext.models.openai import OpenAIChatCompletionClient

# Logger.
logging.getLogger(TRACE_LOGGER_NAME).setLevel(logging.INFO)
logger = logging.getLogger(EVENT_LOGGER_NAME)
logger.setLevel(logging.INFO)

# Tracing setup.
jaeger_exporter = OTLPSpanExporter(endpoint="http://localhost:4317", insecure=True)
tracer_provider = TracerProvider(
    resource=Resource({"service.name": "autogen-test-agentchat"})
)
span_processor = BatchSpanProcessor(jaeger_exporter)
tracer_provider.add_span_processor(span_processor)
tracer = tracer_provider.get_tracer(__name__)

# Note: This example uses mock tools instead of real APIs for demonstration purposes
def search_web_tool(query: str) -> str:
    if "2006-2007" in query:
        return """Here are the total points scored by Miami Heat players in the 2006-2007 season:
        Udonis Haslem: 844 points
        Dwayne Wade: 1397 points
        James Posey: 550 points
        ...
        """
    elif "2007-2008" in query:
        return "The number of total rebounds for Dwayne Wade in the Miami Heat season 2007-2008 is 214."
    elif "2008-2009" in query:
        return "The number of total rebounds for Dwayne Wade in the Miami Heat season 2008-2009 is 398."
    return "No data found."


def percentage_change_tool(start: float, end: float) -> float:
    return ((end - start) / start) * 100


async def main():
    model_client = OpenAIChatCompletionClient(model="gpt-4o")

    planning_agent = AssistantAgent(
        "PlanningAgent",
        description="An agent for planning tasks, this agent should be the first to engage when given a new task.",
        model_client=model_client,
        system_message="""
        You are a planning agent.
        Your job is to break down complex tasks into smaller, manageable subtasks.
        Your team members are:
            WebSearchAgent: Searches for information
            DataAnalystAgent: Performs calculations

        You only plan and delegate tasks - you do not execute them yourself.

        When assigning tasks, use this format:
        1. <agent> : <task>

        After all tasks are complete, summarize the findings and end with "TERMINATE".
        """,
    )

    web_search_agent = AssistantAgent(
        "WebSearchAgent",
        description="An agent for searching information on the web.",
        tools=[search_web_tool],
        model_client=model_client,
        system_message="""
        You are a web search agent.
        Your only tool is search_tool - use it to find information.
        You make only one search call at a time.
        Once you have the results, you never do calculations based on them.
        """,
    )

    data_analyst_agent = AssistantAgent(
        "DataAnalystAgent",
        description="An agent for performing calculations.",
        model_client=model_client,
        tools=[percentage_change_tool],
        system_message="""
        You are a data analyst.
        Given the tasks you have been assigned, you should analyze the data and provide results using the tools provided.
        If you have not seen the data, ask for it.
        """,
    )

    text_mention_termination = TextMentionTermination("TERMINATE")
    max_messages_termination = MaxMessageTermination(max_messages=25)
    termination = text_mention_termination | max_messages_termination

    selector_prompt = """Select an agent to perform task.

    {roles}

    Current conversation context:
    {history}

    Read the above conversation, then select an agent from {participants} to perform the next task.
    Make sure the planner agent has assigned tasks before other agents start working.
    Only select one agent.
    """

    task = "Who was the Miami Heat player with the highest points in the 2006-2007 season, and what was the percentage change in his total rebounds between the 2007-2008 and 2008-2009 seasons?"
    runtime = SingleThreadedAgentRuntime(tracer_provider=tracer_provider)
    with tracer.start_as_current_span("runtime"):
        runtime.start()
        team = SelectorGroupChat(
            [planning_agent, web_search_agent, data_analyst_agent],
            model_client=model_client,
            termination_condition=termination,
            selector_prompt=selector_prompt,
            allow_repeated_speaker=True,  # Allow an agent to speak multiple turns in a row.
            runtime=runtime,
        )
        await Console(team.run_stream(task=task))
    
    await runtime.close()
    await model_client.close()

asyncio.run(main())
```

```
---------- user ----------
Who was the Miami Heat player with the highest points in the 2006-2007 season, and what was the percentage change in his total rebounds between the 2007-2008 and 2008-2009 seasons?
---------- PlanningAgent ----------
To address your query, we'll undertake the following steps:

1. Search for the Miami Heat player with the highest points in the 2006-2007 season.
2. Find the season statistics, specifically the total rebounds, for this player in the 2007-2008 and 2008-2009 seasons.
3. Calculate the percentage change in total rebounds between these two seasons.

Tasks:
1. WebSearchAgent: Identify the Miami Heat player with the highest points in the 2006-2007 NBA season.
2. WebSearchAgent: Find total rebounds for the identified player for the 2007-2008 NBA season.
3. WebSearchAgent: Find total rebounds for the same player for the 2008-2009 NBA season.
4. DataAnalystAgent: Calculate the percentage change in total rebounds from the 2007-2008 season to the 2008-2009 season for the identified player.
---------- WebSearchAgent ----------
[FunctionCall(id='call_v6JnRSbsMagqMeBFES1EYd8M', arguments='{"query":"Miami Heat highest points scorer 2006-2007 season"}', name='search_web_tool')]
---------- WebSearchAgent ----------
[FunctionExecutionResult(content='Here are the total points scored by Miami Heat players in the 2006-2007 season:\n        Udonis Haslem: 844 points\n        Dwayne Wade: 1397 points\n        James Posey: 550 points\n        ...\n        ', name='search_web_tool', call_id='call_v6JnRSbsMagqMeBFES1EYd8M', is_error=False)]
---------- WebSearchAgent ----------
Here are the total points scored by Miami Heat players in the 2006-2007 season:
        Udonis Haslem: 844 points
        Dwayne Wade: 1397 points
        James Posey: 550 points
        ...
        
---------- WebSearchAgent ----------
Dwayne Wade was the Miami Heat player with the highest points in the 2006-2007 season, scoring 1,397 points.

Next, let's find Dwyane Wade's total rebounds for the 2007-2008 and 2008-2009 NBA seasons.
---------- WebSearchAgent ----------
[FunctionCall(id='call_h83aEeUb2pskMiqJAu5kI4IC', arguments='{"query": "Dwyane Wade total rebounds 2007-2008 season"}', name='search_web_tool'), FunctionCall(id='call_mVvoSXQqACQ8hO9IcXEp3eMt', arguments='{"query": "Dwyane Wade total rebounds 2008-2009 season"}', name='search_web_tool')]
---------- WebSearchAgent ----------
[FunctionExecutionResult(content='The number of total rebounds for Dwayne Wade in the Miami Heat season 2007-2008 is 214.', name='search_web_tool', call_id='call_h83aEeUb2pskMiqJAu5kI4IC', is_error=False), FunctionExecutionResult(content='The number of total rebounds for Dwayne Wade in the Miami Heat season 2008-2009 is 398.', name='search_web_tool', call_id='call_mVvoSXQqACQ8hO9IcXEp3eMt', is_error=False)]
---------- WebSearchAgent ----------
The number of total rebounds for Dwayne Wade in the Miami Heat season 2007-2008 is 214.
The number of total rebounds for Dwayne Wade in the Miami Heat season 2008-2009 is 398.
---------- DataAnalystAgent ----------
[FunctionCall(id='call_qnkB04Cjq7BegJBDiXj0BF2g', arguments='{"start":214,"end":398}', name='percentage_change_tool')]
---------- DataAnalystAgent ----------
[FunctionExecutionResult(content='85.98130841121495', name='percentage_change_tool', call_id='call_qnkB04Cjq7BegJBDiXj0BF2g', is_error=False)]
---------- DataAnalystAgent ----------
85.98130841121495
---------- PlanningAgent ----------
Summary of Findings:
- The Miami Heat player with the highest points in the 2006-2007 season was Dwayne Wade, with 1,397 points.
- Dwayne Wade's total rebounds for the 2007-2008 season were 214.
- For the 2008-2009 season, his total rebounds were 398.
- The percentage change in Dwayne Wade's total rebounds from the 2007-2008 season to the 2008-2009 season was approximately 86.0%.

TERMINATE
```

<img width="1270" alt="image" src="https://github.com/user-attachments/assets/7d8fbee1-f7e5-4f99-86ad-bd9371617bd8" />

heyitsaamir reviewed (approved):
Great find with the links!

TomeHirata commented:
Hi, @ekzhu. Thank you for fixing the span structure! Could you also address another point I raised in the ticket?

>Moreover, most spans only have messaging.destination and messaging.operation as their attributes, which is not enough for debugging and improving the system. At least we would like to see the input and output of each message handling.

The lack of meaningful attributes makes it difficult to use the Otel functionality of autogen-core. As I mentioned in the ticket, it would be great if we could include input and output for each agent call at least. 
Let me know if I should create a separate ticket. cc: @BenWilson2

ekzhu commented:
Thanks for the advice. I will create a separate issue and we can continue our discussion there.
